### PR TITLE
NEW Add table to store rules transformation for some fields (like auto upper/lower case)

### PR DESCRIPTION
**Pull request by atm-aurelien**
NEW: dictionnary to change case of given fields (uppercase, ucwords, RegEx...)
This was asked from a few customers to get forced uppercased third-party names and cities.
A const DICT_FIELD_CASES activates or not the case change.

New table on DB:
CREATE TABLE IF NOT EXISTS `llx_c_field_cases` (
`rowid` int(11) NOT NULL,
  `table_name` varchar(50) NOT NULL,
  `field_name` varchar(50) NOT NULL,
  `transformation` varchar(100) NOT NULL,
  `active` tinyint(1) NOT NULL DEFAULT '1'
) ENGINE=InnoDB AUTO_INCREMENT=5 DEFAULT CHARSET=utf8;
ALTER TABLE `llx_c_field_cases`
 ADD PRIMARY KEY (`rowid`);

### CODE REVIEW

File: htdocs/admin/dict.php [hunk 1]
```diff
@@ -634,6 +650,11 @@
         {
             setEventMessage($langs->transnoentities("RecordSaved"));
         	$_POST=array('id'=>$id);	// Clean $_POST array, we keep only
+        	
+        	//Exécution d'un trigger
```
Comment (rdoursenaud) [inline htdocs/admin/dict.php:654]: English comments please

File: htdocs/admin/dict.php [hunk 2]
```diff
@@ -684,6 +713,13 @@
         {
             setEventMessage($db->error(),'errors');
         }
+		else 
+		{
+			//Exécution d'un trigger
```
Comment (rdoursenaud) [inline htdocs/admin/dict.php:718]: English comments please

File: htdocs/conf/conf.php.example [hunk 3]
```diff
@@ -1,326 +0,0 @@
-<?php
```
Comment (rdoursenaud) [inline htdocs/conf/conf.php.example:1]: There's no reason for removing this file

File: htdocs/core/class/html.formmail.class.php.orig [hunk 4]
```diff
@@ -0,0 +1,930 @@
+<?php
```
Comment (rdoursenaud) [inline htdocs/core/class/html.formmail.class.php.orig:1]: This .orig file has nothing to do here

File: htdocs/langs/fr_FR/admin.lang [hunk 5]
```diff
@@ -822,6 +822,7 @@ DictionaryUnits=Unités
 DictionaryProspectStatus=Statuts de prospection
 DictionaryHolidayTypes=Type de congès
 DictionaryOpportunityStatus=Statut d'opportunités pour les affaires/projets
+DictionaryFieldCases=Mise en forme forcée de certains champs (majuscules, camelCase, RegEx...)
```
Comment (rdoursenaud) [inline htdocs/langs/fr_FR/admin.lang:825]: French translations should be contributed at [Transifex](https://www.transifex.com/dolibarr-association/dolibarr/dashboard/) not appear in commits.
On the other hand en_US are required here.

File: htdocs/core/lib/functions.lib.php [hunk 6]
```diff
@@ -38,6 +38,59 @@
...
+					}
+					break;
+			}
+		}
+	}
+}
+
+/**
+ * UCWords with defined separators
+ * (e.g.: Chateau De L'atour-maubourg --> Château De L'Atour-Maubourg)
```
Comment (rdoursenaud) [inline htdocs/core/lib/functions.lib.php:71]: English example would be better.
On a funny note, I doubt that the code will add the circumflex accent ;)

File: htdocs/core/triggers/interface_99_modDict_Dict.class.php [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+	 * All functions "runTrigger" are triggered if file is inside directory htdocs/core/triggers or htdocs/module/code/triggers (and declared)
+     *
+     * @param string		$action		Event action code
+     * @param Object		$object     Object concerned. Some context information may also be provided into array property object->context.
+     * @param User		    $user       Object user
+     * @param Translate 	$langs      Object langs
+     * @param conf		    $conf       Object conf
+     * @return int         				<0 if KO, 0 if no triggered ran, >0 if OK
+     */
+    public function run_trigger($action, $object, User $user, Translate $langs, Conf $conf)
```
Comment (rdoursenaud) [inline htdocs/core/triggers/interface_99_modDict_Dict.class.php:52]: `run_trigger` is deprecated, please use `runTrigger()`.

File: htdocs/core/lib/functions.lib.php [hunk 8]
```diff
@@ -38,6 +38,59 @@
...
 
 /**
+ * Function to transform field from a dictionnary
+ * 
+ * @param 	$db	Database 	Object
+ * @param	$string			String to transform
+ * @param	$tableName		Table of record
+ * @param	$fieldName		Field name
+ */
+function dol_transform_field(&$db, &$string, $tableName, $fieldName) {
```
Comment (rdoursenaud) [inline htdocs/core/lib/functions.lib.php:48]: Your documentation is missing the type of each fields which should go _before_ the variable name.
eg:

``` php
/**
 * This function does something
 *
 * @param DoliDB $db Database handler
 */
```

File: htdocs/core/lib/admin.lib.php [hunk 9]
```diff
@@ -860,6 +860,23 @@ function unActivateModule($value, $requiredby=1)
...
 }
 
+/**
+ * Find value from dictionnary
+ * 
+ *  @param	 	string 				$dictionnary	Dictionnary
+ *  @param	 	array 				$filter	  		Filter
+ *  @return  	array           	$line
+ */ 
+function dolibarr_find_dictionnary(&$db, $dictionnary, $filter) {
```
Comment (rdoursenaud) [inline htdocs/core/lib/admin.lib.php:870]: The parameters documentation does not match the function declaration.

**Comment by rdoursenaud** [comment]
The new table should be added to `htdocs/install/mysql/tables/llx_c_field_cases.sql` and to the 3.9 migration at `htdocs/install/mysql/migration/3.8.0-3.9.0.sql`

**Comment by eldy** [comment]
For better visibility, we must limit trigger to CRUD action on dolibarr business object.
Dictionnaries are not business object bu a technical setup. Using the  dol_transform_field  directly into the dictionary page would be enough.
Also renaming table llx_c_field_cases into llx_c_field_rules would be better. We would be able to extend this table for other transformation rules or surface controls later.

**Comment by aljawaid** [comment]
This would prove useful for things such as postage labels (extrafields used to save price wording on price stickers per product) and cities (Royal Mail UK suggests city in uppercase format on address labels so better for copy/paste situations.

**Comment by eldy** [comment]
I close request because 6.0 introduce table llx_default_value that can manage default sort order, default field value, default filter and can be used to set in next version default field for focus or default transformation rules. So this feature should be implemented in a future using lx_default_value and the new menu that appears in v6 "default values"

**Pull request closed without merging**